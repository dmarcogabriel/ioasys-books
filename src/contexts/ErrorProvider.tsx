import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
