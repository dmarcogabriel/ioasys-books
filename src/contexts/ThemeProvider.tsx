import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
