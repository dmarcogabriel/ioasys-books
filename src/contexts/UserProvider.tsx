import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
