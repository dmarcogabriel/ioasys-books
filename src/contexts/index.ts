export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
