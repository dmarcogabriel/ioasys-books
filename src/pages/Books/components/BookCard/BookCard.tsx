import React from 'react';
import {TouchableWithoutFeedback, View} from 'react-native';
import {Card} from 'react-native-elements';
import {Book} from '../../../../@types/book.interface';
import {
  BookCardContent,
  BookCardImage,
  BookTitle,
  BookAuthor,
  BookInfo,
  BookInfoText,
} from './styles';

interface BookCardProps {
  book: Book;
  onPress: (book: any) => void;
}

export const BookCard = ({book, onPress}: BookCardProps) => {
  const handlePressBook = () => onPress(book);

  return (
    <TouchableWithoutFeedback onPress={handlePressBook}>
      <Card>
        <BookCardContent>
          <BookCardImage source={{uri: book.imageUrl}} resizeMode="contain" />
          <View>
            <BookTitle>{book.title}</BookTitle>
            {book.authors.map((author, i) => (
              <BookAuthor key={`${i}`}>{author}</BookAuthor>
            ))}
            <BookInfo>
              <BookInfoText>{`${book.pageCount} PÃ¡ginas`}</BookInfoText>
              <BookInfoText>{book.publisher}</BookInfoText>
              <BookInfoText>{`Publicado em ${book.published}`}</BookInfoText>
            </BookInfo>
          </View>
        </BookCardContent>
      </Card>
    </TouchableWithoutFeedback>
  );
};
