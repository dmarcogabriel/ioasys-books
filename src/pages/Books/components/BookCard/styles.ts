import styled from 'styled-components/native';
import {Text} from 'react-native-elements';
import {ThemeProps} from '../../../../@types/theme.interface';

export const BookCardContent = styled.View`
  flex-direction: row;
`;

export const BookCardImage = styled.Image`
  width: 81px;
  height: 122px;
  margin-right: 16px;
`;

export const BookTitle = styled(Text)`
  font-size: 14px;
  color: ${({theme}: ThemeProps) => theme.colors.dark};
  font-weight: bold;
`;

export const BookAuthor = styled(Text)`
  font-size: 12px;
  color: ${({theme}: ThemeProps) => theme.colors.primary};
`;

export const BookInfo = styled.View`
  margin-top: 8px;
`;

export const BookInfoText = styled(Text)`
  font-size: 12px;
  color: ${({theme}: ThemeProps) => theme.colors.grey};
`;
