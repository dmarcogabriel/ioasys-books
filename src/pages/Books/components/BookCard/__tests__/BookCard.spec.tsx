import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {BookCard, BookCardProps} from '../BookCard';
import {ThemeProvider} from '../../../../../contexts';
import {Book} from '../../../../../@types/book.interface';
import renderer from 'react-test-renderer';

const ComponentWrapper = (props: BookCardProps) => (
  <ThemeProvider>
    <BookCard {...props} />
  </ThemeProvider>
);
const mockBook: Book = {
  title: 'Test Book',
  authors: ['Author 1', 'Author 2'],
  pageCount: 123,
  publisher: 'Tester',
  published: 1992,
  imageUrl: 'test',
  id: 'test',
  description: 'Testing',
  category: 'Test',
  isbn10: 'test-10',
  isbn13: 'test-13',
  language: 'test',
};
let mockOnPressResult: Book | null = null;
const mockOnPress = jest.fn((book: Book) => {
  mockOnPressResult = book;
});

describe('Books/components/BookCard', () => {
  it('should render correctly', () => {
    const {getByTestId} = render(
      <ComponentWrapper book={mockBook} onPress={mockOnPress} />,
    );
    const bookImage = getByTestId('bookImage');
    const bookTitle = getByTestId('bookTitle');
    const bookAuthor1 = getByTestId('bookAuthor1');
    const bookAuthor2 = getByTestId('bookAuthor2');
    const bookPageCount = getByTestId('bookPageCount');
    const bookPublished = getByTestId('bookPublished');
    const bookPublisher = getByTestId('bookPublisher');

    expect(bookImage).toHaveProp('source', {uri: 'test'});
    expect(bookTitle.children[0]).toBe('Test Book');
    expect(bookAuthor1.children[0]).toBe('Author 1');
    expect(bookAuthor2.children[0]).toBe('Author 2');
    expect(bookPageCount.children[0]).toBe('123 PÃ¡ginas');
    expect(bookPublished.children[0]).toBe('Publicado em 1992');
    expect(bookPublisher.children[0]).toBe('Tester');
  });

  it('should get Book ID by clicking', () => {
    const {getByTestId} = render(
      <ComponentWrapper book={mockBook} onPress={mockOnPress} />,
    );
    const bookContainerButton = getByTestId('bookContainerButton');

    fireEvent.press(bookContainerButton);

    expect(mockOnPress).toHaveBeenCalled();
    expect(mockOnPressResult).toBeTruthy();
    if (mockOnPressResult) {
      expect(mockOnPressResult.id).toBe('test');
      expect(mockOnPressResult.title).toBe('Test Book');
      expect(mockOnPressResult.authors[0]).toBe('Author 1');
      expect(mockOnPressResult.authors[1]).toBe('Author 2');
      expect(mockOnPressResult.pageCount).toBe(123);
      expect(mockOnPressResult.publisher).toBe('Tester');
      expect(mockOnPressResult.published).toBe(1992);
      expect(mockOnPressResult.imageUrl).toBe('test');
    }
  });

  it('should match snapshot', () => {
    expect(
      renderer
        .create(<ComponentWrapper book={mockBook} onPress={mockOnPress} />)
        .toJSON(),
    ).toMatchSnapshot();
  });
});
