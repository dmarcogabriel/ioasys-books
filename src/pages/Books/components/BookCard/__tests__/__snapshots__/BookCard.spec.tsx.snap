// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books/components/BookCard should match snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="bookContainerButton"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
      ]
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": "test",
        }
      }
      style={
        Array [
          Object {
            "height": 122,
            "marginRight": 16,
            "width": 81,
          },
        ]
      }
      testID="bookImage"
    />
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#333333",
              "fontSize": 14,
              "fontWeight": "bold",
            },
          ]
        }
        testID="bookTitle"
      >
        Test Book
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#B22E6F",
              "fontSize": 12,
            },
          ]
        }
        testID="bookAuthor1"
      >
        Author 1
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#B22E6F",
              "fontSize": 12,
            },
          ]
        }
        testID="bookAuthor2"
      >
        Author 2
      </Text>
      <View
        style={
          Array [
            Object {
              "marginTop": 8,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 12,
              },
            ]
          }
          testID="bookPageCount"
        >
          123 PÃ¡ginas
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 12,
              },
            ]
          }
          testID="bookPublisher"
        >
          Tester
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#999999",
                "fontSize": 12,
              },
            ]
          }
          testID="bookPublished"
        >
          Publicado em 1992
        </Text>
      </View>
    </View>
  </View>
</View>
`;
